src/shredder/reed_solomon.rs

```diff
@@ -51,48 +48,65 @@ pub(super) fn reed_solomon_shred(
     let coding_parts = rs::encode(num_data, num_coding, payload.chunks(shred_bytes)).unwrap();
 
     // map raw data/coding parts to `DataShred` and `CodingShred`
-    let payload_from_index_and_data = |index_in_slice: usize, data: Vec<u8>| ShredPayload {
+    let payload_from_index_and_data = |shred_index: ShredIndex, data: Vec<u8>| ShredPayload {
         header: header.clone(),
-        index_in_slice,
+        shred_index,
         data: data.into(),
     };
+
+    // unwrap below should be safe because of the assertion above
     let data_shreds = payload
         .chunks(shred_bytes)
         .enumerate()
-        .map(|(i, p)| DataShred(payload_from_index_and_data(i, p.to_vec())))
+        .map(|(i, p)| {
+            DataShred(payload_from_index_and_data(
+                ShredIndex::new(i).unwrap(),
+                p.to_vec(),
+            ))
+        })
         .collect();
+    // unwrap below should be safe because of the assertion above
     let coding_shreds = coding_parts
         .into_iter()
         .enumerate()
-        .map(|(i, p)| CodingShred(payload_from_index_and_data(i, p)))
+        .map(|(i, p)| CodingShred(payload_from_index_and_data(ShredIndex::new(i).unwrap(), p)))
         .collect();
 
     Ok((data_shreds, coding_shreds))
 }
 
 /// Reconstructs the raw data from the given shreds.
+///
+/// Errors
+///
+/// If fewer than DATA_SHREDS elements in `shreds` are Some() then returns Err(ReedSolomonDeshredError::NotEnoughShreds).
+/// If the restored payload is larger than [`MAX_DATA_PER_SLICE_AFTER_PADDING`] then returns Err(ReedSolomonDeshredError::TooMuchData).
 pub(super) fn reed_solomon_deshred(
-    shreds: &[ValidatedShred],
+    shreds: &[Option<ValidatedShred>; TOTAL_SHREDS],
     num_data: usize,
     num_coding: usize,
     coding_offset: usize,
 ) -> Result<Vec<u8>, ReedSolomonDeshredError> {
     assert!(coding_offset <= DATA_SHREDS);
-    if shreds.len() < DATA_SHREDS {
+    let shreds_cnt = shreds.iter().filter(|s| s.is_some()).count();
+    if shreds_cnt < DATA_SHREDS {
         return Err(ReedSolomonDeshredError::NotEnoughShreds);
     }
-    if shreds.len() > TOTAL_SHREDS {
-        return Err(ReedSolomonDeshredError::TooManyShreds);
-    }
+    assert!(shreds.len() <= TOTAL_SHREDS);
 
     // filter to split data and coding shreds
-    let data = shreds.iter().filter_map(|s| match &s.payload_type {
-        ShredPayloadType::Data(d) => Some((d.index_in_slice, &d.data)),
-        ShredPayloadType::Coding(_) => None,
+    let data = shreds.iter().filter_map(|s| {
```

Comment:
makes this easier as well, as we can just look at `shreds[..coding_offset]` and assert that they have type `ShredPayloadType::Data`, respectively below look at `shreds[coding_offset..]` and assert that they have type `ShredPayloadType::Coding`